compiler/src/iree/compiler/Dialect/HAL/Utils/LLVMCodeGenUtils.cpp

```diff
@@ -0,0 +1,51 @@
+// Copyright 2025 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Dialect/HAL/Utils/LLVMCodeGenUtils.h"
+
+#include "llvm/MC/MCSubtargetInfo.h"
+#include "llvm/Target/TargetMachine.h"
+
+namespace mlir::iree_compiler::IREE::HAL {
+
+void populateLLVMFuncTargetAttrs(ModuleOp moduleOp,
+                                 const llvm::TargetMachine &targetMachine,
+                                 mlir::MLIRContext *context) {
+  if (!context) {
+    context = moduleOp.getContext();
+  }
+
+  llvm::MCSubtargetInfo const *subTargetInfo =
+      targetMachine.getMCSubtargetInfo();
+
+  const std::vector<llvm::SubtargetFeatureKV> enabledFeatures =
+      subTargetInfo->getEnabledProcessorFeatures();
+
+  auto plussedFeatures = llvm::to_vector(
+      llvm::map_range(enabledFeatures, [](llvm::SubtargetFeatureKV feature) {
+        return std::string("+") + feature.Key;
+      }));
+
+  auto plussedFeaturesRefs = llvm::to_vector(llvm::map_range(
+      plussedFeatures, [](auto &it) { return StringRef(it.c_str()); }));
+
+  auto fullTargetFeaturesAttr =
+      LLVM::TargetFeaturesAttr::get(context, plussedFeaturesRefs);
+
+  StringRef targetCPU = targetMachine.getTargetCPU();
+
+  Block &bodyBlock = moduleOp.getBodyRegion().front();
+  for (auto funcOp : bodyBlock.getOps<LLVM::LLVMFuncOp>()) {
+    if (!funcOp.getTargetFeatures().has_value()) {
+      funcOp.setTargetFeaturesAttr(fullTargetFeaturesAttr);
+    }
+    if (!funcOp.getTargetCpu().has_value() && !targetCPU.empty()) {
+      funcOp.setTargetCpuAttr(StringAttr::get(context, targetCPU));
+    }
+  }
+}
```

Comment:
style nit: I'd remove the blank lines. We don't add blank line right after each statement.

> Use vertical whitespace sparingly; unnecessary blank lines make it harder to see overall code structure. Use blank lines only where they aid the reader in understanding the structure.

https://google.github.io/styleguide/cppguide.html#Vertical_Whitespace

Reply:
Good point, I've removed most of the newlines in this block.